File: domain/src/main/java/ch/threema/domain/protocol/api/work/WorkDirectoryContact.java
package ch.threema.domain.protocol.api.work;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class WorkDirectoryContact extends WorkContact {
    @Nullable
    public final String csi;
    public final List<String> categoryIds = new ArrayList<>();
    public final WorkOrganization organization = new WorkOrganization();

    public WorkDirectoryContact(
        @NonNull String threemaId,
        @NonNull byte[] publicKey,
        @Nullable String firstName,
        @Nullable String lastName,
        @Nullable String csi,
        @Nullable String jobTitle,
        @Nullable String department
    ) {
        super(threemaId, publicKey, firstName, lastName, jobTitle, department);
        this.csi = csi;
    }

    public String getInitial(boolean sortByFirstName) {
        String name;
        if (sortByFirstName) {
            name = (firstName != null ? firstName + " " : "") +
                (lastName != null ? lastName : "");

        } else {
            name = (lastName != null ? lastName + " " : "") +
                (firstName != null ? firstName : "");
        }

        if (!name.isEmpty()) {
            return name.substring(0, 1);
        }
        return " ";
    }
}

File: domain/src/main/java/ch/threema/domain/protocol/api/work/WorkDirectoryContact.kt
package ch.threema.domain.protocol.api.work

import java.time.Instant

class WorkDirectoryContact(
    threemaId: String,
    publicKey: ByteArray,
    firstName: String?,
    lastName: String?,
    jobTitle: String?,
    department: String?,
    availability: String?,
    workLastFullSyncAt: Instant?,
    @JvmField val csi: String?,
) : WorkContact(
    threemaId = threemaId,
    publicKey = publicKey,
    firstName = firstName,
    lastName = lastName,
    jobTitle = jobTitle,
    department = department,
    availability = availability,
    workLastFullSyncAt = workLastFullSyncAt,
) {
    @JvmField
    val categoryIds: MutableList<String?> = ArrayList()

    @JvmField
    val organization: WorkOrganization = WorkOrganization()

    fun getInitial(sortByFirstName: Boolean): String {
        val name = if (sortByFirstName) {
            (if (firstName != null) "$firstName " else "") + (lastName ?: "")
        } else {
            (if (lastName != null) "$lastName " else "") + (firstName ?: "")
        }
        if (!name.isEmpty()) {
            return name.substring(0, 1)
        }
        return " "
    }

    override fun toString(): String =
        "WorkDirectoryContact(threemaId=$threemaId, publicKey=${publicKey.contentToString()}, firstName=$firstName, lastName=$lastName, " +
            "jobTitle=$jobTitle, department=$department, availability=$availability, csi=$csi, workLastFullSyncAt=$workLastFullSyncAt)"
}
